README.md
# pysensorpush - Python interface for the SensorPush API

Python implementation of the [SensorPush temperature and humidity/hygrometer sensors](https://www.amazon.com/SensorPush-Wireless-Thermometer-Hygrometer-Android/dp/B01AEQ9X9I?tag=rynoshark-20) cloud API which
supports both temperature (&deg;F) and humidity (Rh) sensors. 

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=WREP29UDAMB6G)

NOTE: Ideally SensorPush sensors would be located within range of a [SensorPush G1 WiFi Gateway](https://www.amazon.com/SensorPush-G1-WiFi-Gateway-Anywhere/dp/B01N17RWWV?tag=rynoshark-20) for continously collecting and publishing data from the sensors to the SensorPush cloud. However, SensorPush sensors can also synchronize historical data over Bluetooth when nearby to an iOS or Android device with the SensorPush app.

## Installation

```
pip3 install pysensorpush
```

## Enable the SensorPush API

While the SensorPush API is in beta, you may need to contact [support@sensorpush.com](mailto:support@sensorpush.com?subject=[api]) to get access to the API for your account. Here are details direct from SensorPush:

> If you're ready to use the beta of the SensorPush API, please review the updated Terms of Service at this link on our website (http://www.sensorpush.com/legal/info). 
>
> These Terms are based largely on the Terms of Service already in place. The primary difference is that we added language to cover the API, so if you accept these terms, please respond to support@sensorpush.com with the following:
>
> 1.) Confirmation:  "Yes, I accept these Terms of Service"
> 2.) Your Gateway email address
>
> We need these two things so we can grant you access to the API.
>
> If you have any questions. Please just send them to support@sensorpush.com and include "[api]" in the subject

## Examples

```python
sensorpush = PySensorPush(username, password)

sensorpush.gateways

sensorpush.sensors

sensorpush.samples
```

See also [example-client.py](example-client.py) for a working example.

## See Also

* [SensorPush API Docs](http://www.sensorpush.com/api/docs)
* [SensorPush for Home Assistant](https://github.com/rsnodgrass/hass-sensorpush)

#### Hardware

* [SensorPush Wireless Thermometer/Hygrometer - Humidity & Temperature Smart Sensor](https://www.amazon.com/SensorPush-Wireless-Thermometer-Hygrometer-Android/dp/B01AEQ9X9I?tag=rynoshark-20)
* [SensorPush G1 WiFi Gateway](https://www.amazon.com/SensorPush-G1-WiFi-Gateway-Anywhere/dp/B01N17RWWV?tag=rynoshark-20)
* [SensorPush](https://sensorpush.com) (official product page)

## Future Enhancements

* improve error handling/edge conditions

* allow fetching data directly from the sensor via Bluetooth (no cloud dependency required)


CHANGELOG.md
## 0.1.2 (2019-08-4)

* automatically re-authenticate when necessary (rather than relying on client checking)

## 0.1.1 (2019-08-31)

* added update() method to trigger update of cached gateways and devices

## 0.1.0 (2019-08-30)

* initial release of Python SensorPush API


example-client.py
#!/usr/local/bin/python3

import os
import sys
import pprint
import logging

from pysensorpush import PySensorPush 

def setup_logger():
    logger = logging.getLogger()
    logger.setLevel(logging.DEBUG)

    handler = logging.StreamHandler(sys.stdout)
    handler.setLevel(logging.DEBUG)

    formatter = logging.Formatter('%(name)s - %(levelname)s - %(message)s')
    handler.setFormatter(formatter)
    logger.addHandler(handler)

def main():
    user = os.getenv('SENSORPUSH_USER', None)
    password = os.getenv('SENSORPUSH_PASSWORD', None)

    if (user == None) or (password == None):
        print("ERROR! Must define env variables SENSORPUSH_USER and SENSORPUSH_PASSWORD")
        raise SystemExit

    #setup_logger()
    pp = pprint.PrettyPrinter(indent = 2)
 
    sensorpush = PySensorPush(user, password)

    print('--Gateways--')
    pp.pprint(sensorpush.gateways)

    print("\n--Sensors--")
    pp.pprint(sensorpush.sensors)

    print("\n--Samples--")
    pp.pprint(sensorpush.samples)

if __name__ == "__main__":
    main()


pysensorpush/2.py
"""
A Python wrapper for the Foobot air quality sensor API.
Allows for data to be queried in three ways:
  1. Get the latest sample from the foobot
  2. Get data from an interval
  3. Get data from a time span
"""

import requests

__version__ = '0.2'
BASE_URL = 'https://api.Foobot.io/v2'


class Foobot:
    """Class for authentication and getting foobot devices."""

    def __init__(self, apikey, username, password):
        """Authenticate the username and password."""
        self.username = username
        self.password = password
        self.session = requests.Session()

        self.auth_header = {'Accept': 'application/json;charset=UTF-8',
                            'X-API-KEY-TOKEN': apikey}

        token = self.login()
        if token is None:
            raise ValueError("Provided username or password is not valid.")

        self.auth_header['X-AUTH-TOKEN'] = token

    def login(self):
        """Log into a foobot device."""
        url = '{base}/user/{user}/login/'.format(base=BASE_URL,
                                                 user=self.username)
        req = self.session.get(url,
                               auth=(self.username, self.password),
                               headers=self.auth_header)
        return req.headers['X-AUTH-TOKEN'] if req.text == "true" else None

    def devices(self):
        """Get list of foobot devices owned by logged in user."""
        url = '{base}/owner/{user}/device/'.format(base=BASE_URL,
                                                   user=self.username)
        req = self.session.get(url, headers=self.auth_header)

        def create_device(device):
            """Helper to create a FoobotDevice based on a dictionary."""
            return FoobotDevice(auth_header=self.auth_header,
                                user_id=device['userId'],
                                uuid=device['uuid'],
                                name=device['name'],
                                mac=device['mac'])

        return [create_device(device) for device in req.json()]


# pylint: disable=too-many-arguments
class FoobotDevice:
    """Represents a foobot device."""

    def __init__(self, auth_header, user_id, uuid, name, mac):
        """Create a foobot device instance used for getting data samples."""
        self.auth_header = auth_header
        self.user_id = user_id
        self.uuid = uuid
        self.name = name
        self.mac = mac
        self.session = requests.Session()

    def latest(self):
        """Get latest sample from foobot device."""
        url = '{base}/device/{uuid}/datapoint/{period}/last/{sampling}/'
        url = url.format(base=BASE_URL,
                         uuid=self.uuid,
                         period=0,
                         sampling=0)
        req = self.session.get(url, headers=self.auth_header)
        return req.json()

    def data_period(self, period, sampling):
        """Get a specified period of data samples."""
        url = '{base}/device/{uuid}/datapoint/{period}/last/{sampling}/'
        url = url.format(base=BASE_URL,
                         uuid=self.uuid,
                         period=period,
                         sampling=sampling)

        req = self.session.get(url, headers=self.auth_header)
        return req.json()

    def data_range(self, start, end, sampling):
        """Get a specified range of data samples."""
        url = '{base}/device/{uuid}/datapoint/{start}/{end}/{sampling}/'
        url = url.format(base=BASE_URL,
                         uuid=self.uuid,
                         start=start,
                         end=end,
                         sampling=sampling)

        req = self.session.get(url, headers=self.auth_header)
        return req.json()


pysensorpush/sensor.py
"""Python class for SensorPush sensor."""

import logging
from pysensorpush.const import (
    QUERY_SAMPLES_ENDPOINT
)

LOG = logging.getLogger(__name__)

def assert_is_dict(var):
    """Assert variable is from the type dictionary."""
    if var is None or not isinstance(var, dict):
        return {}
    return var


class SPSensor(object):
    """SensorPush sensor implementation."""

    def __init__(self, name, attrs, sensorpush_session):
        """Initialize SensorPush sensor object.
        :param name: sensor name
        :param attrs: sensor attributes
        :param sensorpush_session: PySensorPush shared session
        """
        self.name = name
        self._attrs = attrs
        self._session = sensorpush_session

        # make sure self._attrs is a dict
        self._attrs = assert_is_dict(self._attrs)

    def __repr__(self):
        """Representation string of object."""
        return "<{0}: {1}>".format(self.__class__.__name__, self.name)

    @property
    def attrs(self):
        """Return sensor attributes."""
        return self._attrs

    @attrs.setter
    def attrs(self, value):
        """Override sensors attributes."""
        self._attrs = value

    @property
    def temperature(self):
        """Returns temperature from latest sample."""
        return None

    @property
    def humidity(self):
        """Returns humididy from latest sample."""
        return None

    def update(self):
        """Update object properties."""
        self._attrs = self._session.refresh_attributes(self.name)
        self._attrs = assert_is_dict(self._attrs)

        # force associated gateway to update
        if self.gateway:
            self.gateway.update()


pysensorpush/__init2__.py
# coding: utf-8
"""Base Python Class file for Netgear Arlo camera module."""
import logging
import requests

from pyarlo.base_station import ArloBaseStation
from pyarlo.camera import ArloCamera
from pyarlo.media import ArloMediaLibrary
from pyarlo.const import (
    BILLING_ENDPOINT, DEVICES_ENDPOINT,
    FRIENDS_ENDPOINT, LOGIN_ENDPOINT, PROFILE_ENDPOINT,
    PRELOAD_DAYS, RESET_ENDPOINT)

_LOGGER = logging.getLogger(__name__)


class PyArlo(object):
    """Base object for Netgar Arlo camera."""

    def __init__(self, username=None, password=None,
                 preload=True, days=PRELOAD_DAYS):
        """Create a PyArlo object.
        :param username: Arlo user email
        :param password: Arlo user password
        :param preload: Boolean to preload video library.
        :param days: If preload, number of days to lookup.
        :returns PyArlo base object
        """
        self.authenticated = None
        self.country_code = None
        self.date_created = None
        self.userid = None
        self.__token = None
        self.__headers = None
        self.__params = None

        self._all_devices = {}

        # set username and password
        self.__password = password
        self.__username = username
        self.session = requests.Session()

        # login user
        self.login()

        # pylint: disable=invalid-name
        self.ArloMediaLibrary = ArloMediaLibrary(self,
                                                 preload=preload,
                                                 days=days)

    def __repr__(self):
        """Object representation."""
        return "<{0}: {1}>".format(self.__class__.__name__, self.userid)

    def login(self):
        """Login to the Arlo account."""
        _LOGGER.debug("Creating Arlo session")
        self._authenticate()

    def _authenticate(self):
        """Authenticate user and generate token."""
        self.cleanup_headers()
        url = LOGIN_ENDPOINT
        data = self.query(
            url,
            method='POST',
            extra_params={
                'email': self.__username,
                'password': self.__password
            })

        if isinstance(data, dict) and data.get('success'):
            data = data.get('data')
            self.authenticated = data.get('authenticated')
            self.country_code = data.get('countryCode')
            self.date_created = data.get('dateCreated')
            self.__token = data.get('token')
            self.userid = data.get('userId')

            # update header with the generated token
            self.__headers['Authorization'] = self.__token

    def cleanup_headers(self):
        """Reset the headers and params."""
        headers = {'Content-Type': 'application/json'}
        headers['Authorization'] = self.__token
        self.__headers = headers
        self.__params = {}

    def query(self,
              url,
              method='GET',
              extra_params=None,
              extra_headers=None,
              retry=3,
              raw=False,
              stream=False):
        """
        Return a JSON object or raw session.
        :param url:  Arlo API URL
        :param method: Specify the method GET, POST or PUT. Default is GET.
        :param extra_params: Dictionary to be appended on request.body
        :param extra_headers: Dictionary to be apppended on request.headers
        :param retry: Attempts to retry a query. Default is 3.
        :param raw: Boolean if query() will return request object instead JSON.
        :param stream: Boolean if query() will return a stream object.
        """
        response = None
        loop = 0

        # always make sure the headers and params are clean
        self.cleanup_headers()

        while loop <= retry:

            # override request.body or request.headers dictionary
            if extra_params:
                params = self.__params
                params.update(extra_params)
            else:
                params = self.__params
            _LOGGER.debug("Params: %s", params)

            if extra_headers:
                headers = self.__headers
                headers.update(extra_headers)
            else:
                headers = self.__headers
            _LOGGER.debug("Headers: %s", headers)

            _LOGGER.debug("Querying %s on attempt: %s/%s", url, loop, retry)
            loop += 1

            # define connection method
            req = None

            if method == 'GET':
                req = self.session.get(url, headers=headers, stream=stream)
            elif method == 'PUT':
                req = self.session.put(url, json=params, headers=headers)
            elif method == 'POST':
                req = self.session.post(url, json=params, headers=headers)

            if req and (req.status_code == 200):
                if raw:
                    _LOGGER.debug("Required raw object.")
                    response = req
                else:
                    response = req.json()

                # leave if everything worked fine
                break

        return response

    @property
    def cameras(self):
        """Return all cameras linked on Arlo account."""
        return self.devices.get('cameras')

    @property
    def base_stations(self):
        """Return all base stations linked on Arlo account."""
        return self.devices.get('base_station')

    @property
    def devices(self):
        """Return all devices on Arlo account."""
        if self._all_devices:
            return self._all_devices

        self._all_devices = {}
        self._all_devices['cameras'] = []
        self._all_devices['base_station'] = []

        url = DEVICES_ENDPOINT
        data = self.query(url)

        for device in data.get('data'):
            name = device.get('deviceName')
            if ((device.get('deviceType') == 'camera' or
                 device.get('deviceType') == 'arloq' or
                 device.get('deviceType') == 'arloqs') and
                    device.get('state') == 'provisioned'):
                camera = ArloCamera(name, device, self)
                self._all_devices['cameras'].append(camera)

            if (device.get('state') == 'provisioned' and
                    (device.get('deviceType') == 'basestation' or
                     device.get('modelId') == 'ABC1000')):
                base = ArloBaseStation(name, device, self.__token, self)
                self._all_devices['base_station'].append(base)

        return self._all_devices

    def lookup_camera_by_id(self, device_id):
        """Return camera object by device_id."""
        camera = list(filter(
            lambda cam: cam.device_id == device_id, self.cameras))[0]
        if camera:
            return camera
        return None

    def refresh_attributes(self, name):
        """Refresh attributes from a given Arlo object."""
        url = DEVICES_ENDPOINT
        response = self.query(url)

        if not response or not isinstance(response, dict):
            return None

        for device in response.get('data'):
            if device.get('deviceName') == name:
                return device
        return None

    @property
    def unseen_videos_reset(self):
        """Reset the unseen videos counter for all cameras."""
        return self.query(RESET_ENDPOINT).get('success')

    @property
    def billing_information(self):
        """Return billing json."""
        url = BILLING_ENDPOINT
        return self.query(url)

    @property
    def shared_users(self):
        """Return shared users json."""
        url = FRIENDS_ENDPOINT
        return self.query(url)

    @property
    def profile(self):
        """Return user profile json."""
        url = PROFILE_ENDPOINT
        return self.query(url)

    @property
    def is_connected(self):
        """Connection status of client with Arlo system."""
        return bool(self.authenticated)

    def update(self, update_cameras=False, update_base_station=False):
        """Refresh object."""
        self._authenticate()

        # update attributes in all cameras to avoid duped queries
        if update_cameras:
            url = DEVICES_ENDPOINT
            response = self.query(url)
            if not response or not isinstance(response, dict):
                return

            for camera in self.cameras:
                for dev_info in response.get('data'):
                    if dev_info.get('deviceName') == camera.name:
                        _LOGGER.debug("Refreshing %s attributes", camera.name)
                        camera.attrs = dev_info

                # preload cached videos
                # the user is still able to force a new query by
                # calling the Arlo.video()
                camera.make_video_cache()

        # force update base_station
        if update_base_station:
            for base in self.base_stations:
                base.update()


pysensorpush/__init__.py
"""Base Python Class file for SensorPush."""

import time
import logging
import requests

from pysensorpush.sensor import SPSensor
from pysensorpush.const import ( OAUTH_AUTHORIZE_ENDPOINT,
                                 OAUTH_TOKEN_ENDPOINT, 
                                 LIST_SENSORS_ENDPOINT,
                                 LIST_GATEWAYS_ENDPOINT,
                                 QUERY_SAMPLES_ENDPOINT )

LOG = logging.getLogger(__name__)

ACCESS_TOKEN_EXPIRY=30
REFRESH_TOKEN_EXPIRY=60

class PySensorPush(object):
    """Base object for SensorPush."""

    def __init__(self, username=None, password=None):
        """Create a PySensorPush object.
        :param username: SensorPush user email
        :param password: SensorPush user password
        :returns PySensorPush base object
        """
        self.__session = requests.Session()

        self.__token = None
        self.__headers = None
        self.__params = None

        # login the user
        self.__username = username
        self.__password = password
        self.login()

        self._all_gateways = None
        self._all_sensors = None

    def __repr__(self):
        """Object representation."""
        return "<{0}: {1}>".format(self.__class__.__name__, self.__username)

    def login(self):
        """Login to the SensorPush account and generate access token"""
        self.reset_headers()

        # authenticate with user/password
        data = self.query(
            OAUTH_AUTHORIZE_ENDPOINT,
            extra_params={
                'email':    self.__username,
                'password': self.__password
            },
            force_login=False)

        self.__apikey = data.get('apikey')
        self.__authorization = data.get('authorization')

        # get OAuth access token
        data = self.query(
            OAUTH_TOKEN_ENDPOINT,
            extra_params={
                'authorization': self.__authorization
            },
            force_login=False)
        self.__token = data.get('accesstoken')
        self.__refresh_token = data.get('refreshtoken')
        self.__token_timestamp = time.time()

    def _is_access_token_expired(self):
        return (time.time() - self.__token_timestamp / 60) > (ACCESS_TOKEN_EXPIRY - 1)

    @property
    def is_connected(self):
        """Connection status of client with SensorPush cloud service."""
        return bool(self.__token) and not self._is_access_token_expired()

    def reset_headers(self):
        """Reset the headers and params."""
        self.__headers = {
            'User-Agent':    'pysensorpush (https://github.com/rsnodgrass/pysensorpush)',
            'accept':        'application/json',
            'Content-Type':  'application/json',
            'Authorization':  self.__token
        }
        self.__params = {}

    def query(self, url, method='POST', extra_params=None, extra_headers=None, retry=3, force_login=True):
        """
        Returns a JSON object for an HTTP request.
        :param url: API URL
        :param method: Specify the method GET, POST or PUT (default=POST)
        :param extra_params: Dictionary to be appended on request.body
        :param extra_headers: Dictionary to be apppended on request.headers
        :param retry: Retry attempts for the query (default=3)
        """
        response = None
        self.reset_headers() # ensure the headers and params are reset to the bare minimum

        # FIXME: this should really use refresh token, if possible, to reauthenticate
        if force_login and not self.is_connected:
            self.login()

        loop = 0
        while loop <= retry:

            # override request.body or request.headers dictionary
            params = self.__params
            if extra_params:
                params.update(extra_params)
            LOG.debug("Params: %s", params)

            headers = self.__headers
            if extra_headers:
                headers.update(extra_headers)
            LOG.debug("Headers: %s", headers)

            loop += 1
            LOG.debug("Querying %s on attempt: %s/%s", url, loop, retry)

            # define connection method
            request = None
            if method == 'GET':
                request = self.__session.get(url, headers=headers)
            elif method == 'PUT':
                request = self.__session.put(url, headers=headers, json=params)
            elif method == 'POST':
                request = self.__session.post(url, headers=headers, json=params)
            else:
                LOG.error("Invalid request method '%s'", method)
                return None

            if request and (request.status_code == 200):
                response = request.json()
                break # success!

        return response

    @property
    def sensors(self):
        """Return all sensors registered with the SensorPush account."""
        if self._all_sensors:
            return self._all_sensors

        self._all_sensors = self.query(LIST_SENSORS_ENDPOINT)
        LOG.debug("Response sensors = %s", self._all_sensors)
        return self._all_sensors

    @property
    def gateways(self):
        """Return all gateways registered with the SensorPush account."""
        if self._all_gateways:
            return self._all_gateways

        self._all_gateways = self.query(LIST_GATEWAYS_ENDPOINT)
        LOG.debug("Response gateways = %s", self._all_gateways)
        return self._all_gateways

    @property
    def samples(self, limit=1, startTime=None, stopTime=None):
        """Return samples from the SensorPush account.
        :param limit:     how many samples to return, up to 20 (default=1)
        :param startTime: start timestamp range with this format YYYY-MM-DDThh:mm:ss.000Z
        :param stopTime:  stop timestamp range with this format YYYY-MM-DDThh:mm:ss.000Z"""

        params = { 'limit': limit }
        if startTime:
            params['startTime'] = startTime
        if stopTime:
            params['stopTime'] = stopTime

        result = self.query(QUERY_SAMPLES_ENDPOINT, extra_params=params)
        LOG.debug("Samples (limit %d) = %s", limit, result)
        return result

    def update(self, update_gateways=False, update_sensors=False):
        """Refresh any cached state."""
        self.login()

        if update_gateways:
            # clear cache and force update
            self._all_gateways = None
            force_update = self.gateways

        if update_sensors:
            # clear cache and force update
            self._all_sensors = None
            force_update = self.sensors
